File: Data Structures & Algorithms/copy-linked-list-with-random-pointer/submission-2.kt
/*
// Definition for a Node.
class Node(var `val`: Int) {
    var next: Node? = null
    var random: Node? = null
}
*/

class Solution {
    fun copyRandomList(head: Node?): Node? {

        var hash = HashMap<Node?, Node?>() // old: new

        fun clone(node: Node?): Node? {
            if (node == null) {
                return null
            }
            if (hash[node] != null) {
                return hash[node]!!
            }
            val newNode = Node(node!!.`val`)
            hash[node] = newNode
            newNode.next = clone(node?.next)
            newNode.random = clone(node?.random)
            return newNode
        }

        return clone(head)

    }
}

File: Data Structures & Algorithms/reorder-linked-list/submission-2.swift
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     var val: Int
 *     var next: ListNode?
 *     init(_ val: Int) {
 *         self.val = val
 *         self.next = nil
 *     }
 * }
 */

class Solution {
    func reorderList(_ head: ListNode?) {

        func reverse(_ node: ListNode?) -> ListNode? {
            if node == nil {
                return node
            }
            var prev: ListNode? = nil
            var current = node
            while current != nil {
                let temp = current?.next
                current?.next = prev
                prev = current
                current = temp
            }
            return prev
        }

        var slow = head
        var fast = head
        while fast != nil && fast?.next != nil {
            slow = slow?.next
            fast = fast?.next?.next
        }

        var left = head
        var right = reverse(slow)

        while left !== right {
            var leftNext = left?.next
            var rightNext = right?.next

            left?.next = right
            right?.next = leftNext

            left = leftNext
            right = rightNext
        }



    }
}
